pyo3-derive-backend/src/module.rs

```diff
@@ -191,9 +192,25 @@ pub fn add_fn_to_module(
     let python_name = &spec.python_name;
 
     let wrapper = function_c_wrapper(&func.sig.ident, &spec);
+    let mut maybe_module_arg = None;
+    let mod_pointer;
+    let name;
 
+    if is_free {
+        mod_pointer = quote!(let mod_ptr = std::ptr::null_mut(););
+        name = quote!(let name = std::ptr::null_mut(););
+    } else {
+        maybe_module_arg = Some(quote!(, module: &pyo3::types::PyModule));
+        mod_pointer = quote!(let mod_ptr = <::pyo3::types::PyModule as ::pyo3::conversion::AsPyPointer>::as_ptr(module););
+        name = quote!(
+            let name = unsafe { pyo3::ffi::PyModule_GetNameObject(mod_ptr) };
+            if name.is_null() {
+                //TODO Error handling
+            }
+        );
+    }
     Ok(quote! {
-        fn #function_wrapper_ident(py: pyo3::Python) -> pyo3::PyObject {
+        fn #function_wrapper_ident(py: pyo3::Python #maybe_module_arg) -> pyo3::PyObject {
```

Comment:
How about here, instead of requiring the module arg optionally depending on the `free` attribute, we add a little trait to `derive_utils`, e.g.

```rust
trait WrapPyFunctionArguments<'a> {
    fn arguments(self) -> (Python<'a>, Option<&'a PyModule>);
}

impl<'a> WrapPyFunctionArguments<'a> for Python<'a> {
    fn arguments(self) -> (Python<'a>, Option<&'a PyModule>>) { (self, None) }
}

impl<'a> WrapPyFunctionArguments<'a> for &'a PyModule {
    fn arguments(self) -> (Python<'a>, Option<&'a PyModule>>) { (self.py(), Some(self)) }
}
```

Then it should be possible to call either `wrap_pyfunction!(foo)(py)` or `wrap_pyfunction!(foo)(module)` if we change this line to:

```suggestion
        fn #function_wrapper_ident(args: impl pyo3::derive_utils::WrapPyFunctionArguments) -> pyo3::PyObject {
```

And in this generated wrapper code here we could do different things depending on whether a `PyModule` is provided.

Reply:
EDIT: This idea _might not_ work because `wrap_pyfunction!` takes a reference to the wrapper, and by adding the `impl Trait` argument I don't think it can have a reference to it.

If that's the case, I'd still like us to look for a similar trick so that the new functionality comes with new syntax, without changing the existing code.